core/backend/src/IModelDb.ts

```diff
@@ -2356,31 +2409,96 @@ export class BriefcaseDb extends IModelDb {
       this.concurrencyControl.onSavedChanges();
   }
 
-  private async lockSchema(requestContext: AuthorizedClientRequestContext): Promise<void> {
+  private static async lockSchema(requestContext: AuthorizedClientRequestContext, iModelId: GuidString, changeSetId: GuidString, briefcaseId: number): Promise<void> {
     requestContext.enter();
     const lock = new Lock();
-    lock.briefcaseId = this.briefcaseId;
+    lock.briefcaseId = briefcaseId;
     lock.lockLevel = LockLevel.Exclusive;
     lock.lockType = LockType.Schemas;
     lock.objectId = "0x1";
-    lock.releasedWithChangeSet = this._changeSetId;
-    lock.seedFileId = this.iModelId;
+    lock.releasedWithChangeSet = changeSetId;
+    lock.seedFileId = iModelId;
 
     Logger.logTrace(loggerCategory, `lockSchema`);
-    const res = await BriefcaseManager.imodelClient.locks.update(requestContext, this.iModelId, [lock]);
+    const res = await BriefcaseManager.imodelClient.locks.update(requestContext, iModelId, [lock]);
     if (res.length !== 1 || res[0].lockLevel !== LockLevel.Exclusive)
-      throw new IModelError(IModelStatus.UpgradeFailed, "Could not acquire schema lock", Logger.logError, loggerCategory, () => this.getRpcProps());
+      throw new IModelError(IModelStatus.UpgradeFailed, "Could not acquire schema lock", Logger.logError, loggerCategory, () => ({ iModelId, changeSetId, briefcaseId }));
+  }
+
+  /**
+   * Upgrades the profile or domain schemas and returns the new change set id
+   */
+  private static async upgradeProfileOrDomainSchemas(requestContext: AuthorizedClientRequestContext, briefcaseProps: LocalBriefcaseProps, upgradeOptions: UpgradeOptions, changeSetDescription: string): Promise<GuidString> {
+    requestContext.enter();
+
+    await this.lockSchema(requestContext, briefcaseProps.iModelId, briefcaseProps.changeSetId, briefcaseProps.briefcaseId);
+    requestContext.enter();
+
+    const openMode = OpenMode.ReadWrite;
+    const nativeDb = this.openDgnDb({ path: briefcaseProps.fileName }, openMode, upgradeOptions);
+    assert(!nativeDb.hasUnsavedChanges(), "Expected schema upgrade to have saved any changes made");
+    const changeSetId = nativeDb.getParentChangeSetId();
+    if (!nativeDb.hasPendingTxns())
+      return changeSetId;
+    const reversedChangeSetId = nativeDb.getReversedChangeSetId();
+    assert(reversedChangeSetId === undefined, "Expected schema upgrade to have failed if there were reversed changes in the briefcase");
+
+    const token: IModelRpcProps = {
+      key: "",

```

Comment:
use `IModelRpcOpenProps` and get rid of key:

Reply:
i made several changes to ensure you don't return without closing the iModel. Please review.